src/data-model-providers/codegen/CodegenDataModelProvider.cpp

```diff
@@ -553,165 +420,143 @@ std::optional<unsigned> CodegenDataModelProvider::TryFindEndpointIndex(EndpointI
     return std::make_optional<unsigned>(idx);
 }
 
-DataModel::EndpointEntry CodegenDataModelProvider::NextEndpoint(EndpointId before)
-{
-    std::optional<unsigned> before_idx = TryFindEndpointIndex(before);
-    if (!before_idx.has_value())
-    {
-        return DataModel::EndpointEntry::kInvalid;
-    }
-    return FirstEndpointEntry(*before_idx + 1, mEndpointIterationHint);
-}
-
-DataModel::ClusterEntry CodegenDataModelProvider::FirstServerCluster(EndpointId endpointId)
+DataModel::MetadataList<DataModel::ServerClusterEntry> CodegenDataModelProvider::ServerClusters(EndpointId endpointId)
 {
     const EmberAfEndpointType * endpoint = emberAfFindEndpointType(endpointId);
-    VerifyOrReturnValue(endpoint != nullptr, DataModel::ClusterEntry::kInvalid);
-    VerifyOrReturnValue(endpoint->clusterCount > 0, DataModel::ClusterEntry::kInvalid);
-    VerifyOrReturnValue(endpoint->cluster != nullptr, DataModel::ClusterEntry::kInvalid);
 
-    return FirstServerClusterEntry(endpointId, endpoint, 0, mServerClusterIterationHint);
-}
+    DataModel::MetadataList<DataModel::ServerClusterEntry> result;
 
-std::optional<unsigned> CodegenDataModelProvider::TryFindClusterIndex(const EmberAfEndpointType * endpoint, ClusterId id,
-                                                                      ClusterSide side) const
-{
-    const unsigned clusterCount = endpoint->clusterCount;
-    unsigned hint               = side == ClusterSide::kServer ? mServerClusterIterationHint : mClientClusterIterationHint;
+    VerifyOrReturnValue(endpoint != nullptr, result);
+    VerifyOrReturnValue(endpoint->clusterCount > 0, result);
+    VerifyOrReturnValue(endpoint->cluster != nullptr, result);
+
+    const EmberAfCluster * begin = endpoint->cluster;
+    const EmberAfCluster * end   = endpoint->cluster + endpoint->clusterCount;
 
-    if (hint < clusterCount)
+    size_t serverClusterCount = 0;
+    for (const EmberAfCluster * cluster = begin; cluster != end; cluster++)
     {
-        const EmberAfCluster & cluster = endpoint->cluster[hint];
-        if (((side == ClusterSide::kServer) && cluster.IsServer()) || ((side == ClusterSide::kClient) && cluster.IsClient()))
+        if (cluster->IsServer())
         {
-            if (cluster.clusterId == id)
-            {
-                return std::make_optional(hint);
-            }
+            serverClusterCount++;
```

Comment:
Is there a reason we are re-inventing `emberAfClusterCountForEndpointType(endpoint, /* server = */ true)`?